### PR TITLE
Fix Issue 24135 - Eponymous template member overloads not shown as call candidates

### PR DESCRIPTION
**Pull request by ntrel**

### CODE REVIEW

File: compiler/src/dmd/frontend.h [hunk 1]
```diff
@@ -6840,7 +6840,7 @@ class TemplateDeclaration final : public ScopeDsymbol
     const char* kind() const override;
     const char* toChars() const override;
     const char* toCharsNoConstraints() const;
-    const char* toCharsMaybeConstraints(bool includeConstraints) const;
+    const char* toCharsFull(bool includeConstraints, bool showOneMember = true) const;
```
Comment (ibuclaw) [inline compiler/src/dmd/frontend.h:6843]: Check for this in `template.h`

**Review by ibuclaw — DISMISSED** [review]

File: compiler/src/dmd/frontend.h (hunk 1 above)
Comment (ntrel) [reply compiler/src/dmd/frontend.h]: toCharsMaybeConstraints isn't in template.h. I had forgotten to rename the first parameter, now done.

File: compiler/src/dmd/func.d [hunk 2]
```diff
@@ -3518,10 +3518,24 @@ if (is(Decl == TemplateDeclaration) || is(Decl == FuncDeclaration))
...
                                 : int.max;
     const(char)* constraintsTip;
-    // determine if the first candidate was printed
-    int printed;
 
-    bool matchSymbol(Dsymbol s, bool print, bool single_candidate = false)
+    int printed = 0; // number of candidates printed
+    int count = 0; // total candidates
+    bool child; // true if inside an eponymous template
+    string prefix()
```
Comment (maxhaton) [inline compiler/src/dmd/func.d:3525]: More descriptive name / comment please.

Comment (ntrel) [reply compiler/src/dmd/func.d]: Updated

**Comment by ntrel** [comment]
@maxhaton Ping

File: compiler/src/dmd/frontend.h (hunk 1 above)
Comment (ntrel) [reply compiler/src/dmd/frontend.h]: Ping @ibuclaw please remove changes requested

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
I would also add a test with two eponymous templates that are in the same overload set but have different template parameters + an invalid call, just to make sure that both eponymous candidates are printed correctly.

Other than that, please rebase this and I will merge it.

File: compiler/src/dmd/hdrgen.d [hunk 3]
```diff
@@ -1974,11 +1975,13 @@ void toCharsMaybeConstraints(const TemplateDeclaration td, ref OutBuffer buf, re
     }
     buf.writeByte(')');
 
-    if (td.onemember)
+    if (hgs.showOneMember && td.onemember)
     {
         if (const fd = td.onemember.isFuncDeclaration())
+        //~ const FuncDeclaration fd = td.onemember.isFuncDeclaration();
+        //~ if (fd && fd.type)
```
Comment (ntrel) [inline compiler/src/dmd/hdrgen.d:1982]:
```suggestion
```

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**